### PR TITLE
feat: add Emission contract tests

### PR DESCRIPTION
**Pull request by nvtaveras**
### Description

This adds tests for the `calculateEmission()` and `emitTokens()` functions in the Emission contract according to the [test plan](https://www.notion.so/mentolabs/Governance-Test-Plan-4c38412bebe74095b025b26d2dc34ac0#f1ac9f6f82f04fa7b5c111cbb5fb85be). The numbers were checked agains't the jupyter notebook as well as the unit tests written in mento-core.

There are some very minor differences with that notebook outputs due to precision, mainly because the notebook doesn't use Decimal so some precision is lost, and also the value of the A constant is `454968308.0947435` while in the contract implementation we only use the integer part.

```
========================
Running Emission Contract tests on network with chain id: 44787
========================

    emission.calculateEmission()
      ✔ should calculate the correct amount after 1 month (815ms)
      ✔ should calculate the correct amount after 6 months (536ms)
      ✔ should calculate the correct amount after 1 year (546ms)
      ✔ should calculate the correct amount after 10 years (551ms)
      ✔ should calculate the correct amount after 15 years (568ms)
      ✔ should calculate the correct amount after 25 years (553ms)
      ✔ should calculate the correct amount after 30 years (533ms)
      ✔ should calculate the correct amount after 40 years (551ms)
    emission.emitTokens()
      ✔ should emit the correct amount after 1 month (1230ms)
      ✔ should emit the correct amount after 6 months (1212ms)
      ✔ should emit the correct amount after 1 year (1231ms)
      ✔ should emit the correct amount after 10 years (1260ms)
      ✔ should emit the correct amount after 15 years (1201ms)
      ✔ should emit the correct amount after 25 years (1337ms)
      ✔ should emit the correct amount after 30 years (1953ms)
      ✔ should emit the correct amount after 40 years (1746ms)
```

### Related issues

- Fixes https://github.com/mento-protocol/governance-tests/issues/5

### CODE REVIEW

File: test/Emission.test.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+describe('Emission Contract', function () {
+  const { provider } = ethers;
+
+  const DAY = 60 * 60 * 24;
+  const MONTH = 30 * DAY;
+  const YEAR = 365 * DAY;
+
+  const TOTAL_EMISSION_SUPPLY = BigInt('650000000000000000000000000');
+  const EMISSION_SCHEDULE: [string, number, BigInt][] = [
```
Comment (baroooo) [inline test/Emission.test.ts:21]: linter

File: test/Emission.test.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        );
+
+        const emittedSoFar = await emission.totalEmittedAmount();
+        expect(await mentoToken.emittedAmount()).to.equal(emittedSoFar);
+
+        const [signer] = await ethers.getSigners();
+        await emission.connect(signer!).emitTokens();
+        expect(BigInt(await helpers.time.latest()) - start).to.eq(timeToTravel);
+
+        expect(await emission.totalEmittedAmount()).to.equal(
```
Comment (baroooo) [inline test/Emission.test.ts:120]: NIT: lets stick to .eq or .equal

File: test/Emission.test.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    await helpers.reset(hre.network.config.forking.url);
+  });
+
+  before(async function () {
+    const chainId = hre.network.config.chainId;
+    if (!chainId) {
+      throw new Error('Chain ID not found');
+    }
+
+    governanceAddresses = mento.getContractsByChainId(chainId);
```
Comment (baroooo) [inline test/Emission.test.ts:48]: this is not gonna work in the new version of sdk, lets use mento.addresses[chainId]!;

File: test/Emission.test.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+  const YEAR = 365 * DAY;
+
+  const TOTAL_EMISSION_SUPPLY = BigInt('650000000000000000000000000');
+  const EMISSION_SCHEDULE: [string, number, BigInt][] = [
+    ['1 month', MONTH, BigInt('3692586569806444700000000')],
+    ['6 months', 6 * MONTH, BigInt('21843234320275928950000000')],
+    ['1 year', YEAR, BigInt('43528555608969174900000000')],
+    ['10 years', 10 * YEAR, BigInt('325091005879265576600000000')],
+    ['15 years', 15 * YEAR, BigInt('421181077924230273650000000')],
+    ['25 years', 25 * YEAR, BigInt('554584121910167354200000000')],
```
Comment (baroooo) [inline test/Emission.test.ts:27]: we can just add "n" at the end of the numbers instead of casting it to BigInt every time

**Review by baroooo — CHANGES_REQUESTED** [review]

File: test/Emission.test.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    governanceAddresses = mento.getContractsByChainId(chainId);
+    if (!governanceAddresses) {
+      throw new Error('Governance addresses not found for this chain');
+    }
+
+    emission = Emission__factory.connect(
+      governanceAddresses.Emission,
+      // @TODO: Do I need this line below?
```
[the remark this answers is not in the document]
Comment (bayological) [reply test/Emission.test.ts:55]: The issue is related to the differences between versions of ethers being used. The type of provider being used in the test is the [HardhatEthersProvider](https://github.com/NomicFoundation/hardhat/blob/e6a42cf330863a6c64624baedd82fc6c5218a20a/packages/hardhat-ethers/src/internal/hardhat-ethers-provider.ts#L81), which comes from the hardhat-ethers package. This provider is using ethers v6, and it inherits from the Provider class specific to that version. The connect function accepts a Signer or Provider type from ethers v5.7. If you remove the eslint ignore we will get an error as the types don't fully match up due to some functions that were removed on the new Provider. We're lucky that the underlying functions being called by the interactions don't break, so it's fine to overlook this type checking inconsistency for the time being. But if thing's start to look weird then this is the first thing to check. 

To fix this we should either update the mento-types to include types for [ethersv6](https://www.npmjs.com/package/@typechain/ethers-v6) or downgrade this project to use a version of Hardhat that users ethers v5. The former would be easiest IMO but I don't think doing either of these at this point is necessary.

File: test/Emission.test.ts (hunk 2 above)
Comment (nvtaveras) [reply test/Emission.test.ts:120]: I removed the block.timestamp asserts so it should be `.equal` everywhere now

**Comment by nvtaveras** [comment]
@baroooo updated!

**Review by baroooo — APPROVED** [review]
🚢

File: test/Emission.test.ts [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    governanceAddresses = mento.addresses[chainId]!;
+    if (!governanceAddresses) {
+      throw new Error('Governance addresses not found for this chain');
+    }
+
+    emission = Emission__factory.connect(
+      governanceAddresses.Emission,
+      // @TODO: Do I need this line below?
```
Comment (philbow61) [inline test/Emission.test.ts:55]: maybe we can remove the comment after what bayo said

**Review by philbow61 — APPROVED** [review]
Lgtm

**Pull request merged**